src/analysis/processing/qgsalgorithmrasterrank.cpp

```diff
@@ -0,0 +1,309 @@
+
+  auto writer = std::make_unique<QgsRasterFileWriter>( outputFile );
+  writer->setOutputFormat( outputFormat );
+  std::unique_ptr<QgsRasterDataProvider> provider( writer->createMultiBandRaster( outputDataType, cols, rows, outputExtent, outputCrs, mRanks.size() ) );
+  if ( !provider )
+    throw QgsProcessingException( QObject::tr( "Could not create raster output: %1" ).arg( outputFile ) );
+  if ( !provider->isValid() )
+    throw QgsProcessingException( QObject::tr( "Could not create raster output %1: %2" ).arg( outputFile, provider->error().message( QgsErrorMessage::Text ) ) );
+  provider->setNoDataValue( 1, outputNoData );
+
+  std::map<QString, std::unique_ptr<QgsRasterInterface>> inputInterfaces;
+  std::map<QString, std::unique_ptr<QgsRasterBlock>> inputBlocks;
+  for ( QgsMapLayer *layer : mLayers )
+  {
+    QgsRasterLayer *rLayer = qobject_cast<QgsRasterLayer *>( layer );
+    if ( rLayer->crs() != outputCrs )
+    {
+      QgsRasterProjector *projector = new QgsRasterProjector();
+      projector->setCrs( rLayer->crs(), outputCrs, context.transformContext() );
+      projector->setInput( rLayer->dataProvider() );
+      projector->setPrecision( QgsRasterProjector::Exact );
+      inputInterfaces[rLayer->id()].reset( projector );
+    }
+    else
+    {
+      inputInterfaces[rLayer->id()].reset( rLayer->dataProvider()->clone() );
+    }
+    inputBlocks[rLayer->id()] = std::make_unique<QgsRasterBlock>();
+  }
+
+  std::unique_ptr<QgsRasterIterator> rasterIterator = std::make_unique<QgsRasterIterator>( inputInterfaces[mLayers.at( 0 )->id()].get() );
+  rasterIterator->startRasterRead( 1, cols, rows, outputExtent );
+  int rasterIteratorCount = static_cast<int>( rasterIterator->blockCount() );
+
+  std::vector<std::unique_ptr<QgsRasterBlock>> outputBlocks;
+  for ( int i = 0; i < mRanks.size(); i++ )
+  {
+    outputBlocks.push_back( std::make_unique<QgsRasterBlock>() );
+  }
+
+
+  const double step = rasterIteratorCount > 0 ? 100.0 / rasterIteratorCount : 0;
+  for ( int i = 0; i < rasterIteratorCount; i++ )
+  {
+    if ( feedback->isCanceled() )
+    {
+      break;
+    }
+    feedback->setProgress( i * step );
+
+    int iterLeft = 0;
+    int iterTop = 0;
+    int iterCols = 0;
+    int iterRows = 0;
+    QgsRectangle blockExtent;
+    rasterIterator->next( 1, iterCols, iterRows, iterLeft, iterTop, blockExtent );
+
+    for ( const auto &inputInterface : inputInterfaces )
+    {
+      inputBlocks[inputInterface.first].reset( inputInterface.second->block( 1, blockExtent, iterCols, iterRows ) );
+    }
+
+    for ( int i = 0; i < mRanks.size(); i++ )
+    {
+      outputBlocks[i].reset( new QgsRasterBlock( outputDataType, iterCols, iterRows ) );
+      outputBlocks[i]->setNoDataValue( outputNoData );
+    }
+
+    for ( int row = 0; row < iterRows; row++ )
+    {
+      for ( int col = 0; col < iterCols; col++ )
+      {
+        QList<double> values;
```

Comment:
This is allocating memory for every pixel -- we could speed things up (considerably?) by moving this out of the pixel iteration loops, and by using a fixed size std::array instead.